### PR TITLE
core,eth/tracers: don't panic in OnTxEnd if receipt is nil

### PR DESCRIPTION
**Pull request by lightclient**
There are two issues currently with the tracer's `OnTxEnd` invocation in `state_processor`

1. Because it is deferred and uses named return values, the receipt is not yet set in the first return after `ApplyTransactionWithEVM`. `OnTxEnd` expects the receipt to be non-nil, so this causes a panic.
2. ~~The error from `result` is never set into the named return value `err`. So the error reporting for `OnTxEnd` isn't really correct.~~

This PR fixes both of these issues.

cc @s1na

### CODE REVIEW

**Comment by s1na** [comment]
1. Hm I thought it is expected that in case of error the receipt will be null. It can be that the tracers are not handling this well?
2. Yep so here the OnTxEnd captures tx-level errors like insufficient intrinsic gas etc. The "vm errors" will be caught by lower level hooks like OnExit or OnFault.

**Comment by lightclient** [comment]
Okay I see, I updated the PR to just fix 1). Thanks

File: eth/tracers/logger/logger.go [hunk 1]
```diff
@@ -268,7 +268,9 @@ func (l *StructLogger) OnTxEnd(receipt *types.Receipt, err error) {
 		}
 		return
 	}
-	l.usedGas = receipt.GasUsed
+	if receipt != nil {
```
Comment (s1na) [inline eth/tracers/logger/logger.go:271]: I'm surprised you've seen a panic here. As far as I can see in state processor either receipt is nil and error is non-nil (in which case struct logger returns) or there is a receipt.

Comment (lightclient) [reply eth/tracers/logger/logger.go:271]: Ah you're right. The issue issue is actually that `err` is redeclared.

**Review by fjl — APPROVED** [review]

**Comment by lightclient** [comment]
Closing this for now as my change doesn't address the issue and the current code in `state_processor` works. The issue is there are some call sites for `OnTxEnd` which do not pass a receipt, e.g. https://github.com/ethereum/go-ethereum/blob/2d9d42376436cd275c28056cffd0eb97cb8daed8/tests/state_test_util.go#L302

**Pull request closed without merging**